tinygrad/renderer/triton.py

```diff
@@ -0,0 +1,134 @@
+from typing import Dict, List, Final, Callable, DefaultDict
+from collections import defaultdict
+from tinygrad.ops import UnaryOps, BinaryOps, TernaryOps, Op
+from tinygrad.helpers import dtypes, ImageDType, DEBUG, getenv
+from tinygrad.codegen.linearizer import  UOp, UOps
+from tinygrad.shape.symbolic import NumNode
+from triton.compiler import compile as triton_compile
+import hashlib
+import math
+import re
+def next_power_of_2(x):
+  return 1 << (x - 1).bit_length()
+
+def render_valid(node):
+  return '(' * (len(node.nodes) -1) + ') and '.join([f'{n.render()}<{get_max(n+1)}' for n in node.nodes]) if hasattr(node, "nodes") else f"{node.render()}<{get_max(node+1)}"
+
+#NOTE triton requires matching dimensions for load/store, disable this and see TestOps::test_output_padded_conv_transpose2d fail to compile
+def fill_dims_for_idx(idx, dims):
+  return "(" + idx.render() + "+ (" + (f"0 * ({'+'.join(d for d in dims)})))")
+
+def get_max(var):
+  if isinstance(var.max, int): return var.max
+  return re.sub(r'\[(.*?)\]', '', str(var.max))[1:-1]
+
+#NOTE can be removed after https://github.com/gpuocelot/gpuocelot/issues/8 gets resolved
+def remove_single_scalar_curly_braces(ptx_code):
+  return '\n'.join([re.sub(r'\{\s*(%\w+)\s*\}', r'\1', line) for line in ptx_code.split('\n')])
+
+def define_scalar(local_size, triton_type, args):
+  val = (('-' if args<0 else '') + 'float("inf")') if math.isinf(args) else ('float("nan")' if math.isnan(args) else str(args))
+  if len(local_size) > 0:
```

Comment:
Move up in function for early quiting. Also put return on same line

Reply:
wdym by move up in function?

Reply:
Oh nvm messed up viewing it on phone, sorry!